### PR TITLE
tests: posix: pthread: split into pthread_t and pthread_attr_t tests

### LINKED ISSUE #67091: tests: posix: common: rework testsuite to de-duplicate code, narrow test scope, etc
**Is your enhancement proposal related to a problem? Please describe.**
It's become very evident there there is significant bitrot in `tests/posix/common`
* State is spread throughout not only `pthread.c` but most other places

**Describe the solution you'd like**

* Each component should be in it's own `ZTEST_SUITE()` instead of using a single `posix_apis` testsuite
  * They can all still be a part of the `common` app (unless it really makes sense to move them out - e.g. sleep APIs?)
  * barrier, clock, cond, fnmatch, mqueue, mutex, pthread, pthread_attr, rwlock, semaphore, signal, sleep, spinlock, timer, uname
* for `pthread.c`, rather than testing many things in only a couple of `ZTEST()` cases, it's probably better to narrow the of each `ZTEST()` and test many things in many `ZTEST()` cases
* potentially utilize `setup()`, `before()`, `after()`, and `teardown()` callbacks for `ZTEST_SUITE()`
* potentially use `ZTEST_F()` for testcases and fixtures to separately track state for each suite
* make tests more concise
  * avoid use of temporary variables, branches, and printks
  * instead of `ret = pthread_..(); zassert_.. (ret, ...)`, just use `zassert_...(pthread_...())`
  * preserve existing zassert error messages, if it makes sense

**Describe alternatives you've considered**
Not refactoring the testsuite code and having to continue dealing with state being spread everywhere.

**Additional context**
Decided after wrestling with #67085 for days (months?).

Actually, I wanted to put a `struct posix_thread_attr` inside of `struct posix_thread` 6 months ago, and the testsuite / techdebt was the major limiting factor. It only took 6 months of working on this in my "spare time" to make it happen 😅 

- [x] https://github.com/zephyrproject-rtos/zephyr/pull/67094
- [ ] https://github.com/zephyrproject-rtos/zephyr/pull/67575
- [ ] Possibly more?

### PR DESCRIPTION
**Pull request by cfriedt**
**TL;DR - This series of changes is mainly concerned with better testsuite hygiene. It shuffles existing tests around and manages state better with `ZTEST_SUITE()`**

The `posix_apis.pthread_execution` and `posix_apis.pthread_termination` tests were quite unorganized. It was no longer feasible to maintain testsuite state in a coherent way.

This became particularly apparent after work was started on embedding `struct posix_thread_attr` inside of `struct posix_thread`, and using `pthread_attr_t` to maintain thread stack allocation details (See #67178).

The change that embeds `struct posix_thread_attr` inside of `struct posix_thread` will follow shortly.

* make some additional `NULL` checks in `lib/posix/pthread.c`
* create `barrier.c` and move the barrier test to it
* move the `pthread_condattr_t` test to `cond.c`
* move tests from `pthread.c` that operate on `pthread_attr_t` to `pthread_attr.c`
* keep tests in `pthread.c` that operate on `pthread_t` inside of `pthread.c`

(Partially)
Fixes #67091

### CODE REVIEW

File: tests/posix/common/src/pthread_attr.c [hunk 1]
```diff
@@ -0,0 +1,638 @@
...
+
+	/* ensure we can create and join a thread with the default attrs */
+	can_create_thread(&attr);
+
+	/* set stack / addr to the current values of stack / addr */
+	zassert_ok(pthread_attr_setstack(&attr, stackaddr, stacksize));
+	can_create_thread(&attr);
+
+	/* qemu_x86 seems to be unable to set thread stacks to be anything less than 4096 */
+	if (!IS_ENABLED(CONFIG_X86)) {
```
Comment (cfriedt) [inline tests/posix/common/src/pthread_attr.c:346]: @ceolin or perhaps @dcpleung - hoping you can do a review here.

This seems kind of weird to me. The thread stack that is allocated (IIRC via pool) is 4096 bytes (via `K_THREAD_STACK_ARRAY_DEFINE()`). If I subtract 1 from that size, this section breaks.

I know that for statically allocated thread stacks, a smaller size is fine for x86. Does this have to do with MMU / page alignment?

File: tests/posix/common/src/pthread_attr.c [hunk 2]
```diff
@@ -0,0 +1,638 @@
...
+
+	/* set stack / addr to the current values of stack / addr */
+	zassert_ok(pthread_attr_setstacksize(&attr, stacksize));
+	/* ensure we can read back the values we just set */
+	zassert_ok(pthread_attr_getstacksize(&attr, &new_stacksize));
+	zassert_equal(new_stacksize, stacksize);
+	can_create_thread(&attr);
+
+	/* qemu_x86 seems to be unable to set thread stacks to be anything less than 4096 */
+	if (!IS_ENABLED(CONFIG_X86)) {
```
Comment (cfriedt) [inline tests/posix/common/src/pthread_attr.c:427]: Similar case here, @ceolin & @dcpleung

**Comment by cfriedt** [comment]
I guess I could squash the last two commits into one.

File: tests/posix/common/src/pthread.c [hunk 3]
```diff
@@ -574,33 +512,7 @@ static void *create_thread1(void *p1)
 	return NULL;
 }
 
-ZTEST(posix_apis, test_pthread_create_negative)
```
Comment (fabiobaltieri) [inline tests/posix/common/src/pthread.c:599]: hey, sorry I have a hard time following, are these deleted or moved? the commit message suggets that they are moved but the commit only seems to delete

File: tests/posix/common/src/pthread_attr.c [hunk 4]
```diff
@@ -0,0 +1,638 @@
...
+	create_thread_common(NULL, true, true);
+}
+
+ZTEST(posix_apis, test_pthread_attr_static_corner_cases)
+{
+	pthread_attr_t attr1;
+
+	if (IS_ENABLED(CONFIG_DYNAMIC_THREAD)) {
+		ztest_test_skip();
+	}
```
Comment (fabiobaltieri) [inline tests/posix/common/src/pthread_attr.c:94]: you could use `Z_TEST_SKIP_IFDEF` if you really want to

File: tests/posix/common/src/pthread.c (hunk 3 above)
Comment (cfriedt) [reply tests/posix/common/src/pthread.c:599]: They're moved. Probably will be more apparent when I squash the last two commits into one.

Comment (cfriedt) [reply tests/posix/common/src/pthread.c:599]: Fixed!

**Review by fabiobaltieri — DISMISSED** [review]

File: tests/posix/common/src/pthread_attr.c [hunk 5]
```diff
@@ -0,0 +1,636 @@
...
+ZTEST(pthread_attr, test_null_attr)
+{
+	/*
+	 * This test can only succeed when it is possible to call pthread_create() with a NULL
+	 * pthread_attr_t* (I.e. when we have the ability to allocate thread stacks dynamically).
+	 */
+	create_thread_common(NULL, IS_ENABLED(CONFIG_DYNAMIC_THREAD) ? true : false, true);
+}
+
+ZTEST(posix_apis, test_pthread_attr_static_corner_cases)
```
Comment (ycsin) [inline tests/posix/common/src/pthread_attr.c:88]:
```suggestion
ZTEST(pthread_attr, test_pthread_attr_static_corner_cases)
```

Comment (cfriedt) [reply tests/posix/common/src/pthread_attr.c]: Fixed! Good catch

**Comment by cfriedt** [comment]
✅  rebased to fix merge conflict

File: tests/posix/common/src/pthread_attr.c (hunk 4 above)
Comment (cfriedt) [reply tests/posix/common/src/pthread_attr.c]: Fixed!

**Review by ycsin — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]
